src/runtime/js.zig

```diff
@@ -1516,6 +1508,184 @@ pub fn Env(comptime State: type, comptime WebApis: type) type {
+                    .{ .alloc = .if_needed },
+                ) catch unreachable;
+
+                log.debug(.js, "dynamic import", .{
+                    .specifier = specifier_str,
+                    .resource = resource_str,
+                    .referrer_full = referrer_full_url,
+                    .normalized_specifier = normalized_specifier,
+                });
+
+                _dynamicModuleCallback(context, normalized_specifier, &resolver) catch |err| {
+                    log.err(.js, "dynamic module callback", .{
+                        .err = err,
+                    });
+                    // Must be rejected at this point
+                    // otherwise, we will just wait on a pending promise.
+                    std.debug.assert(resolver.getPromise().getState() == .kRejected);
+                };
+                return @constCast(resolver.getPromise().handle);
+            }
+
+            fn _dynamicModuleCallback(
+                context: *JsContext,
+                specifier: []const u8,
+                resolver: *const v8.PromiseResolver,
+            ) !void {
+                const iso = context.isolate;
+                const ctx = context.v8_context;
+
+                const module_loader = context.module_loader;
+                const source = module_loader.func(module_loader.ptr, specifier) catch {
+                    const error_msg = v8.String.initUtf8(iso, "Failed to load module");
+                    _ = resolver.reject(ctx, error_msg.toValue());
+                    return;
+                } orelse {
+                    const error_msg = v8.String.initUtf8(iso, "Module source not available");
+                    _ = resolver.reject(ctx, error_msg.toValue());
+                    return;
+                };
+
+                var try_catch: TryCatch = undefined;
+                try_catch.init(context);
+                defer try_catch.deinit();
+
+                const maybe_promise = context.module(source, specifier, true) catch {
+                    log.err(.js, "module compilation failed", .{
+                        .specifier = specifier,
+                        .exception = try_catch.exception(context.call_arena) catch "unknown error",
+                        .stack = try_catch.stack(context.call_arena) catch null,
+                        .line = try_catch.sourceLineNumber() orelse 0,
+                    });
+                    const error_msg = if (try_catch.hasCaught()) blk: {
+                        const exception_str = try_catch.exception(context.call_arena) catch "Evaluation error";
+                        break :blk v8.String.initUtf8(iso, exception_str orelse "Evaluation error");
+                    } else v8.String.initUtf8(iso, "Module evaluation failed");
+                    _ = resolver.reject(ctx, error_msg.toValue());
+                    return;
+                };
+                const new_module = context.module_cache.get(specifier).?.castToModule();
```

Comment:
I don't see how it's possible, but when I tried this branch with reddit, `context.module_cache.get(specifier)` returns null (and this code crashed). But I don't see how `context.module()` called with `cacheable: true` can return without inserting that entry...

Reply:
The `call_arena` isn't safe to use, e.g.

```zig
 const specifier_str = jsStringToZig(context.call_arena, specifier, iso) catch {
```

The exact lifetime of the call_arena is a bit unspecific, but it's guaranteed to exist for at least the point where a WebAPI function is called (e.g. Window._setTimeout) to the point where data is handed to v8.

Loading a module is _a lot_ wider than that and likely includes many WebApi calls. So imagine you have a module:

```js
// awesome.js
console.log("hello");
// <-- should assume that the call_arena is reset at this point
```
Your specifier (and probably other variables) are created before this code executes, but those allocations will become invaild as soon as `console.log` returns.